KAFKA-15853: Move KafkaConfig log properties and docs out of core

OmniaGM opened the pull request:
part 1 for #15501
### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

OmniaGM committed:
KAFKA-15853: Move KafkaConfig log properties and docs out of core

OmniaGM commented:
@nizhikov can you have a look please?

nizhikov commented on server/src/main/java/org/apache/kafka/server/config/KafkaConfig.java:
```diff
@@ -0,0 +1,146 @@
+import org.apache.kafka.common.config.TopicConfig;
+
+public class KafkaConfig {
+    private final static String LOG_CONFIG_PREFIX = "log.";
```
Minor 

1. Let's rename this to `LOG_PREFIX` to make naming more compact.

2. We have `ServerTopicConfigSynonyms.LOG_PREFIX`, already.
Maybe it will be better to use only one constant?

OmniaGM replied:
Nice catch. I updated the pr to `ServerTopicConfigSynonyms.LOG_PREFIX`

nizhikov commented on build.gradle:
```diff
@@ -1837,6 +1837,8 @@ project(':storage') {
     testImplementation project(':core')
     testImplementation project(':core').sourceSets.test.output
+    testImplementation project(':server')
+    testImplementation project(':server').sourceSets.test.output
```
Tests not required in `storage` module - only main classes. Line 1841 can be removed.

OmniaGM committed:
use ServerTopicConfigSynonyms.LOG_PREFIX

OmniaGM committed:
feedback

mimaison commented on server/src/main/java/org/apache/kafka/server/config/KafkaConfig.java:
```diff
@@ -0,0 +1,146 @@
+public class KafkaConfig {
+
+    /** ********* Log Configuration ***********/
+    public final static String NUM_PARTITIONS_PROP = "num.partitions";
```
For consistency with the other *Configs classes, should we used the `CONFIG` suffix instead of `PROP`?

mimaison replied:
Would it also make sense to colocate the `CONFIG` and `DOC` like we do in the other Config classes?

OmniaGM replied:
> For consistency with the other *Configs classes, should we used the CONFIG suffix instead of PROP?

I thought about it but I wan't sure which would make it easier for people to review. 
We also might need to do the same for LogCleaner configs as well. 

> Would it also make sense to colocate the CONFIG and DOC like we do in the other Config classes?

I considered this but I found it tricky for some docs that refer to more than 1 config in the string interpolation.  For example the `LOG_DIR_DOC` and `LOG_DIRS_DOC` refer each others prop e.g LOG_DIRS_DOC  refer to `LOG_DIR_PROP` and LOG_DIR_DOC  refers to `LOG_DIRS_PROP` same goes to `LOG_ROLL_TIME_HOURS_DOC` and `LOG_ROLL_TIME_MILLIS_DOC`. I thought this why they are grouped at the end after defining the properties names first.

OmniaGM replied:
I just pushed a commit to replace `PROP` with `CONFIG` suffix.

OmniaGM replied:
I reorganised this KafkaLogConfigs in the last commit to have config, doc and default in the same order like the other config classes.

mimaison commented on core/src/main/scala/kafka/server/ZkAdminManager.scala:
```diff
@@ -49,6 +49,7 @@ import org.apache.kafka.common.security.scram.internals.{ScramCredentialUtils, S
 import org.apache.kafka.common.utils.Sanitizer
 import org.apache.kafka.server.common.AdminOperationException
 import org.apache.kafka.server.config.{ConfigEntityName, ConfigType}
+import org.apache.kafka.server.config.KafkaConfig._
```
I wonder if we should just import `KafkaConfig` and use `KAFKA.SOME_CONFIG` instead of importing all/statically KafkaConfig. WDYT?

OmniaGM replied:
<del> For this one and maybe other that import few config from KafkaConfig this would be nice improvement but some classes like DynamicBrokerConfig for example include a large set of config. I can update few to use  `KAFKA.SOME_CONFIG` if they need small set of config and leave the others specially that we when we move them to java we will do this anyway.</del>

Ignore my previous comment I got the comment wrong. At the moment we can't use KafkaConfig everywhere as there's another KafkaConfig in scala

OmniaGM replied:
Also at the moment we can't use `KafkaConfig` everywhere as there's another `KafkaConfig` in scala

mimaison replied:
Ah right, that's a good point. Thanks

OmniaGM replied:
After the last commit, this shouldn't be an issue anymore as the configs are in KafkaLogConfigs instead of KafkaConfig so there is no conflict anymore

mimaison reviewed (commented):
Thanks for the PR. I left a few suggestions.

OmniaGM committed:
replace RROP suffix with CONFIG suffix

OmniaGM committed:
feedback

mimaison commented:
Now that https://github.com/apache/kafka/pull/15075 got merged, this needs rebasing.

OmniaGM commented:
rebased the pr

mimaison commented on server/src/main/java/org/apache/kafka/server/config/KafkaConfig.java:
```diff
@@ -0,0 +1,146 @@
+    public final static String ALTER_CONFIG_POLICY_CLASS_NAME_CONFIG = "alter.config.policy.class.name";
+    public final static String LOG_MESSAGE_DOWN_CONVERSION_ENABLE_CONFIG = ServerTopicConfigSynonyms.serverSynonym(TopicConfig.MESSAGE_DOWNCONVERSION_ENABLE_CONFIG);
+
+    // Document
```
Should it be `Documentation`?

mimaison commented on server/src/main/java/org/apache/kafka/server/config/KafkaConfig.java:
```diff
@@ -0,0 +1,146 @@
+    public final static String MIN_IN_SYNC_REPLICAS_CONFIG = ServerTopicConfigSynonyms.serverSynonym(TopicConfig.MIN_IN_SYNC_REPLICAS_CONFIG);
+    public final static String CREATE_TOPIC_POLICY_CLASS_NAME_CONFIG = "create.topic.policy.class.name";
+    public final static String ALTER_CONFIG_POLICY_CLASS_NAME_CONFIG = "alter.config.policy.class.name";
+    public final static String LOG_MESSAGE_DOWN_CONVERSION_ENABLE_CONFIG = ServerTopicConfigSynonyms.serverSynonym(TopicConfig.MESSAGE_DOWNCONVERSION_ENABLE_CONFIG);
```
Can we use a consistent formatting for down conversion across classes? Either `DOWN_CONVERSION` or `DOWNCONVERSION` but let's pick one.

mimaison reviewed (commented):
Thanks for the updates! I made another pass and left a couple of comments.

fvaleri reviewed (approved):
LGTM, thanks.

showuon commented:
Since Mickael is already reviewing it, I'll focus on other PRs. Thanks @mimaison !

OmniaGM commented:
> @OmniaGM I think this is almost ready, but there are a couple of minor comments to address and some conflicts to fix.

Will be addressing this soon after another pr

OmniaGM committed:
Address mimaison's feedback

OmniaGM commented:
@fvaleri, @nizhikov @mimaison just rebased the pr and addressed the feedback will wait for this one to be merged before rebasing the other pr for KafkaConfig

OmniaGM committed:
fix

mimaison commented:
@OmniaGM Is this the next PR in line for moving broker configs?

OmniaGM committed:
refactor to move the props def into KafkaLogConfigs

OmniaGM commented:
> @OmniaGM Is this the next PR in line for moving broker configs?

Yes, I rebased the PR and also pushed a refactor to follow the same pattern we adopted in https://github.com/apache/kafka/pull/15656#discussion_r1554518847, https://github.com/apache/kafka/pull/15684, https://github.com/apache/kafka/pull/15670

OmniaGM commented:
The last commit renamed the class to `KafkaLogConfigs` instead similar to `KafkaSecurityConfigs` and `ZKConfigs`. I tried to move some to `LogConfig.ServerLogConfig` similar to `LogConfig.RemoteLogConfig` specially that some default configs are defined in `LogConfig`. However I'm bit concern that `LogConfig` seems already huge. What others prefer? Keep it in  `KafkaLogConfigs`  or move them to LogConfig.ServerLogConfig`
cc: @mimaison, @chia7712

chia7712 commented:
> Keep it in KafkaLogConfigs

I prefer this way since `LogConfig` includes too many stuff now ... maybe we should move the constants out of `LogConfig`

chia7712 commented on server/src/main/java/org/apache/kafka/server/config/KafkaLogConfigs.java:
```diff
@@ -0,0 +1,166 @@
+    public final static String LOG_DIRS_CONFIG = LOG_PREFIX + "dirs";
+    public final static String LOG_DIR_CONFIG = LOG_PREFIX + "dir";
+    public static final String LOG_DIR_DEFAULT = "/tmp/kafka-logs";
+    public final static String LOG_DIR_DOC = "The directory in which the log data is kept (supplemental for " + LOG_DIRS_CONFIG + " property)";
```
Should we align the declaration style? I prefer the style: `static final`

OmniaGM committed:
feedback

OmniaGM commented:
> maybe we should move the constants out of `LogConfig`

We can try to do this in a follow up after we merge all PRs for KafkaConfig as am worried about potential circular dependency. And it would better to see the full pic before do improvement for `LogConfig`.

chia7712 commented:
There are build error. please fix it :)

OmniaGM committed:
fix pipeline

OmniaGM committed:
merge trunk

chia7712 commented on server/src/main/java/org/apache/kafka/server/config/KafkaLogConfigs.java:
```diff
@@ -0,0 +1,166 @@
+    public static final String ALTER_CONFIG_POLICY_CLASS_NAME_DOC = "The alter configs policy class that should be used for validation. The class should " +
+            "implement the <code>org.apache.kafka.server.policy.AlterConfigPolicy</code> interface.";
+    public static final String LOG_MESSAGE_DOWNCONVERSION_ENABLE_CONFIG = ServerTopicConfigSynonyms.serverSynonym(TopicConfig.MESSAGE_DOWNCONVERSION_ENABLE_CONFIG);
+    public static final String LOG_MESSAGE_DOWN_CONVERSION_ENABLE_DOC = TopicConfig.MESSAGE_DOWNCONVERSION_ENABLE_DOC;
```
It seems to me we should make consistency for all docs here. 1) make it point to `TopicConfig.xxxDOC` or 2) copy the docs to here.

OmniaGM replied:
I don't think we can copy all configs doc to `TopicConfig` as the docs are different in many places between topic level config and log/topic broker level config. I will copy it for now into KafkaLogConfigs in case the doc diverted in the futrue like the rest of the docs KafkaLogConfigs.

chia7712 reviewed (commented):
@OmniaGM thanks for updated PR. overall +1 and please take a look at one last comment

OmniaGM committed:
duplicate  MESSAGE_DOWNCONVERSION_ENABLE_DOC value to  LOG_MESSAGE_DOWN_CONVERSION_ENABLE_DOC

chia7712 commented:
> hopefully for the last time :)

I feel #15684 will cause many conflicts to this PR, and vice versa :_

OmniaGM commented:
> Let's just let Omnia decides which PR she prefers doing first. Then we can all review that PR, and once merged move onto the next one. This would avoid unnecessary churn as I expect she's getting tired of rebasing all of these PRs every day!

Thanks @mimaison this is really thoughtful! breaking KafkaConfig out of core is one of the hardest cleanup specially if we want to move with smaller/medium PRs. So I don't see any other way than going through it. I think it might be better to take them in order so we can get them merged and the rebases can be easier. 
Let's merge https://github.com/apache/kafka/pull/15684 as
-  it is smaller 
-  more contained than the log config 
- and group configs are more likely to change as part of KIP-848 and any of its followups which make the conflicts much harder and risker to get messed up.

OmniaGM committed:
Merge remote-tracking branch 'apache/trunk' into KAFKA-15853-KafkaConfig-prop-0

OmniaGM committed:
Merge remote-tracking branch 'apache/trunk' into KAFKA-15853-KafkaConfig-prop-0

mimaison commented on core/src/main/scala/kafka/server/ZkAdminManager.scala:
```diff
@@ -79,10 +80,10 @@ class ZkAdminManager(val config: KafkaConfig,
 
   private val createTopicPolicy =
-    Option(config.getConfiguredInstance(KafkaConfig.CreateTopicPolicyClassNameProp, classOf[CreateTopicPolicy]))
+    Option(config.getConfiguredInstance(ALTER_CONFIG_POLICY_CLASS_NAME_CONFIG, classOf[CreateTopicPolicy]))
```
This should be `CREATE_TOPIC_POLICY_CLASS_NAME_CONFIG`

mimaison commented on core/src/test/scala/unit/kafka/log/LogConfigTest.scala:
```diff
@@ -399,18 +400,18 @@ class LogConfigTest {
 
-  /* Verify that when the deprecated config LogMessageTimestampDifferenceMaxMsProp has non default value the new configs
-   * LogMessageTimestampBeforeMaxMsProp and LogMessageTimestampAfterMaxMsProp are not changed from the default we are using
+  /* Verify that when the deprecated config LOG_MESSAGE_TIMESTAMP_DIFFERENCE_MAX_MS_PROP has non default value the new configs
```
`LOG_MESSAGE_TIMESTAMP_DIFFERENCE_MAX_MS_PROP` has been renamed to `LOG_MESSAGE_TIMESTAMP_BEFORE_MAX_MS_CONFIG`
Same below

mimaison commented on core/src/test/scala/unit/kafka/server/KafkaConfigTest.scala:
```diff
@@ -1152,12 +1151,12 @@ class KafkaConfigTest {
+    defaults.setProperty(KafkaLogConfigs.LOG_ROLL_TIME_HOURS_CONFIG, "12")
+    defaults.setProperty(KafkaLogConfigs.LOG_ROLL_TIME_JITTER_HOURS_CONFIG, "11")
+    defaults.setProperty(KafkaLogConfigs.LOG_RETENTION_TIME_HOURS_CONFIG, "10")
+    //For LOG_FLUSH_INTERVAL_MS_PROP
```
This has been renamed to `LOG_FLUSH_INTERVAL_MS_CONFIG`

mimaison reviewed (commented):
Thanks for the updates. I found a few small issues.

OmniaGM committed:
feedback

mimaison commented on core/src/main/scala/kafka/server/ZkAdminManager.scala:
```diff
@@ -80,10 +80,10 @@ class ZkAdminManager(val config: KafkaConfig,
 
   private val alterConfigPolicy =
-    Option(config.getConfiguredInstance(ALTER_CONFIG_POLICY_CLASS_NAME_CONFIG, classOf[AlterConfigPolicy]))
+    Option(config.getConfiguredInstance(CREATE_TOPIC_POLICY_CLASS_NAME_CONFIG, classOf[AlterConfigPolicy]))
```
This one was correct! We need to undo this

OmniaGM replied:
fix it

OmniaGM committed:
fix

mimaison commented on server/src/main/java/org/apache/kafka/server/config/KafkaLogConfigs.java:
```diff
@@ -0,0 +1,170 @@
+    public static final String ALTER_CONFIG_POLICY_CLASS_NAME_CONFIG = "alter.config.policy.class.name";
+    public static final String ALTER_CONFIG_POLICY_CLASS_NAME_DOC = "The alter configs policy class that should be used for validation. The class should " +
+            "implement the <code>org.apache.kafka.server.policy.AlterConfigPolicy</code> interface.";
+    public static final String LOG_MESSAGE_DOWNCONVERSION_ENABLE_CONFIG = ServerTopicConfigSynonyms.serverSynonym(TopicConfig.MESSAGE_DOWNCONVERSION_ENABLE_CONFIG);
```
Nit: There's a mix of `DOWNCONVERSION` and `DOWN_CONVERSION`.

OmniaGM replied:
Good spot. fixed now

mimaison reviewed (approved):
Thanks for all the updates and quickly addressing feedback. LGTM

OmniaGM committed:
rename LOG_MESSAGE_DOWN_CONVERSION_ENABLE_DOC to LOG_MESSAGE_DOWNCONVERSION_ENABLE_DOC

chia7712 commented:
> However I'm bit concern that LogConfig seems already huge. What others prefer? Keep it in KafkaLogConfigs or move them to LogConfig.ServerLogConfig

the most default values of `KafkaLogConfigs` are in `LogConfig`, and they are in different module. That pattern is no similar to `ReplicationConfigs`, `KafkaSecurityConfigs`. Is server-common module more suitable to collect those server side configs since both`storage` and `server` depend on `server-common`. Also, `server-common` has `org.apache.kafka.server.config` package too.

OmniaGM commented:
> > However I'm bit concern that LogConfig seems already huge. What others prefer? Keep it in KafkaLogConfigs or move them to LogConfig.ServerLogConfig
> 
> the most default values of `KafkaLogConfigs` are in `LogConfig`, and they are in different module. That pattern is no similar to `ReplicationConfigs`, `KafkaSecurityConfigs`. Is server-common module more suitable to collect those server side configs since both`storage` and `server` depend on `server-common`. Also, `server-common` has `org.apache.kafka.server.config` package too.

I think it might be better and simpler to move `KafkaLogConfigs` into `server-common` and make LogConfig use them. I'll push an update with this soon

OmniaGM committed:
Move KafkaLogConfigs into server-commons in order to be used by both storage, server and core

chia7712 commented on server-common/src/main/java/org/apache/kafka/server/config/KafkaLogConfigs.java:
```diff
@@ -0,0 +1,188 @@
+ *
+ * Note this is an internal API and subject to change without notice.
+ */
+public class KafkaLogConfigs {
```
Do we need to add prefix "kafka" here? sorry that I try to figure out the naming rule. It seems to me the prefix "kafka" means the class is used to collect other "xxxConfigs" constants and to be the central place for kafka xxx-related configs. For example, `KafkaSecurityConfigs` collects all security-related configs from `BrokerSecurityConfigs`, `SaslConfigs`, `SslConfigs`, etc

OmniaGM replied:
I can rename it to `ServerLogConfigs` or `BrokerLogConfigs` but we already have `LogConfig`

OmniaGM replied:
Just pushed an update. The problem right now that we have the name of the config type without any unnecessary prefixes has been preserved in other modules and am trying to avoid making this pr bigger than it should be. I think at some point we need to revisit some of the existing Config files as some of them breaks few rules like using `PROP` instead of `CONFIG` suffix and use `DEFAULT` as prefix instead of suffix and some of the classes is `Config` instead of `Configs` as well.

chia7712 replied:
> I think at some point we need to revisit some of the existing Config files as some of them breaks few rules like using PROP instead of CONFIG suffix and use DEFAULT as prefix instead of suffix and some of the classes is Config instead of Configs as well.

agree!

chia7712 reviewed (commented):
@OmniaGM thanks for updated PR. overall +1. one last comment left.

OmniaGM committed:
Rename KafkaLogConfigs to ServerLogConfigs

chia7712 reviewed (approved):
LGTM, let us move on

chia7712 commented:
```
[2024-04-18T17:26:16.149Z] [ant:checkstyle] [ERROR] /home/jenkins/workspace/Kafka_kafka-pr_PR-15569/server-common/src/main/java/org/apache/kafka/server/config/ServerLogConfigs.java:22:1: Disallowed import - org.apache.kafka.server.common.MetadataVersion.IBP_3_0_IV1. [ImportControl]

[2024-04-18T17:26:16.149Z] [ant:checkstyle] [ERROR] /home/jenkins/workspace/Kafka_kafka-pr_PR-15569/server-common/src/main/java/org/apache/kafka/server/config/ServerLogConfigs.java:23:1: Disallowed import - org.apache.kafka.server.config.ServerTopicConfigSynonyms.LOG_PREFIX. [ImportControl]

```
there are imports error :_

OmniaGM committed:
fix imports

chia7712 commented:
The failed test `testParseAndValidateAddressesWithReverseLookup` will get fixed by #15758. I will merge this PR
